File: backend/app/models/chapter.py
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey, Text
from sqlalchemy.orm import relationship

from ..database import Base


class Chapter(Base):
    __tablename__ = "chapters"

    id = Column(String(36), primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    content = Column(Text, nullable=False)
    summary = Column(Text, nullable=True)
    word_count = Column(Integer, default=0)
    order = Column(Integer, default=0)
    status = Column(String(50), default="draft")
    novel_id = Column(String(36), ForeignKey("novels.id", ondelete="CASCADE"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    novel = relationship("Novel", back_populates="chapters")

File: frontend/src/services/chapter.ts
import api from "./api";

export type Chapter = {
  id: string;
  title: string;
  content: string;
  summary?: string | null;
  order: number;
  status: string;
  word_count: number;
  novel_id: string;
};

export async function listChapters(novelId: string): Promise<Chapter[]> {
  const { data } = await api.get<Chapter[]>(`/novels/${novelId}/chapters/`);
  return data;
}

export async function createChapter(
  novelId: string,
  payload: Pick<Chapter, "title" | "content"> & Partial<Pick<Chapter, "summary" | "order" | "status">>
): Promise<Chapter> {
  const { data } = await api.post<Chapter>(`/novels/${novelId}/chapters/`, payload);
  return data;
}

export async function updateChapter(
  novelId: string,
  chapterId: string,
  payload: Partial<Pick<Chapter, "title" | "content" | "summary" | "order" | "status">>
): Promise<Chapter> {
  const { data } = await api.put<Chapter>(`/novels/${novelId}/chapters/${chapterId}`, payload);
  return data;
}

export async function deleteChapter(novelId: string, chapterId: string): Promise<void> {
  await api.delete(`/novels/${novelId}/chapters/${chapterId}`);
}

